---
 keepalived/vrrp/vrrp.c |   16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

--- a/keepalived/vrrp/vrrp.c
+++ b/keepalived/vrrp/vrrp.c
@@ -1721,8 +1721,22 @@
 	int ret = 0;
 
 	if (!VRRP_VIP_ISSET(vrrp)) {
+#ifdef _WITH_SNMP_RFCV3_
+		char * reason;
+		if (vrrp->stats->master_reason == VRRPV3_MASTER_REASON_MASTER_NO_RESPONSE)
+			reason = " - master down timer expired";
+		else
+			reason = "";
+#endif
 		log_message(LOG_INFO, "VRRP_Instance(%s) Entering MASTER STATE"
-				    , vrrp->iname);
+#ifdef _WITH_SNMP_RFCV3_
+					"%s"
+#endif
+				    , vrrp->iname
+#ifdef _WITH_SNMP_RFCV3_
+					, reason
+#endif
+					);
 		vrrp_state_become_master(vrrp);
 		ret = 1;
 	} else if (!timer_isnull(vrrp->garp_refresh) &&
